### ISSUE TITLE
Quick guide on how to use ssl

### ISSUE THREAD

**Issue by dmastag** [mean-expert-official/loopback-sdk-builder#459]
#### What type of issue are you creating?
- [ ] Bug
- [ ] Enhancement
- [x] Question

#### What version of this module are you using?
- [ ] 2.0.10 (Stable)
- [x] 2.1.0-rc.13 (2.1 Release Candidate n)
- [ ] Other

Write other if any:

#### Please add a description for your issue:

I updated to rc.13 and as per https://github.com/mean-expert-official/fireloop.io/issues/107#issuecomment-316489886 added

within component-config.json
```
{
  "@mean-expert/loopback-component-realtime": {
  "secure": true
  ....
  }
}
```
within the SDK

```
import {LoopBackConfig} from 'sdk';

LoopBackConfig.setSecureWebSockets();
```

But am still getting 
Error during WebSocket handshake: Unexpected response code: 503

Am I doing something wrong or missed a step?

**Comment by jonathan-casarrubias** [comment]
@dmastag make sure you change your base URL to https or //, instead of http

**Comment by dmastag** [comment]
Already did change to the `private static path: string =` in the `lb.config.ts` file and not working.
Is there anything I have missed?

Am using pm2 with a couple of cores.

**Comment by jonathan-casarrubias** [comment]
@dmastag are you directly serving with loopback or do you have nginx / apache on top?

**Comment by dmastag** [comment]
I am serving using pm2, but I am not sure how the Infrastructure team is setting up domains and https. Is there anything that I should be looking out for?

**Comment by mugbomah** [comment]
@jonathan-casarrubias, I have been battling with this issue as well for close to a week. While websocket works perfectly via http, it does not seem to work with SSL. I have tried the following options:
1. Serving directly with Node over https
2. Serving behind Nginx
Keep getting this error:
> WebSocket connection to 'wss://xxx.com/socket.io/?EIO=3&transport=websocket' failed: Error during WebSocket handshake: Unexpected response code: 502

Standard Loopback API works.

Based on your advise above, I did not set Component-config "secure": true.

Any assistance will be greatly appreciated.

**Comment by jonathan-casarrubias** [comment]
Hey Guys,

Using Nginx with proxy pass is the easier way to make SSL work, I do have a project with 3 different environments running on SSL.

I'm sharing the nginx configuration I'm using

`Note: These are the only 2 steps I'm doing to make it work`

````sh
##
# Redirect all incoming requests to secure connection
##
server {
  listen 80;
  return 301 https://$host$request_uri;
}
##
# Configure SSL proxy pass
##
server {
        listen 443;
        server_name www.my.domain;

        ssl on;
        ssl_certificate      /cd/to/my.crt;
        ssl_certificate_key  /cd/to/my.key;

        location / {
                # First attempt to serve request as file, then
                # as directory, then fall back to displaying a 404.
                proxy_pass http://localhost:3000;
                proxy_http_version 1.1;
                proxy_set_header Upgrade $http_upgrade;
                proxy_set_header Connection 'upgrade';
                proxy_set_header Host $host;
                proxy_cache_bypass $http_upgrade;
        }
}
````

In the Angular CLI environments I have this configuration

````js
export const environment = {
  production: true,
  envName : 'prod',
  apiUrl : '//www.my.domain',
  apiVersion : 'api'
};
````

which is later used when configuring the SDK

`NOTE: Please check I'm using // instead of http or https, this way the SDK will run in both HTTP and HTTPS`

````js
import { environment } from '../environments/environment';

export const BASE_URL = environment.apiUrl ;
export const API_VERSION = environment.apiVersion;
````


Then in my main.ts file

````js
import { LoopBackConfig } from './app/shared/sdk';

LoopBackConfig.setBaseURL(BASE_URL);
LoopBackConfig.setApiVersion(API_VERSION);
````

Other than that I didn't configured nothing in the LoopBack side, I left the configuration like that because locally the projects run in HTTP, once in the server nginx is the one with SSL configuration but then it proxy pass the request to `http://localhost:3000`

And that should be all, after that I'm able to have secure connections

![image](https://user-images.githubusercontent.com/1533239/28542304-57a9af56-7081-11e7-8d3b-37d31da92588.png)

**Comment by mugbomah** [comment]
@jonathan-casarrubias thank you very much! It worked like a charm. 👍 

I had to make 2 changes to my Angular 2 project:
1. Changed environment to `apiUrl : '//www.my.domain',`
2. moved this from app.component.ts to main.ts
```
import { LoopBackConfig } from './app/shared/sdk';
import { environment } from './environments/environment';

LoopBackConfig.setBaseURL(environment.apiUrl);
LoopBackConfig.setApiVersion(environment.apiVersion);
```
Thanks!

**Comment by jonathan-casarrubias** [comment]
Hey @mugbomah thanks for confirming this is working..

I'm adding this to the documentation

Cheers
Jon

**Comment by jonathan-casarrubias** [comment]
I have just added a couple of references in both wiki documentations, for the SDK and within the FireLoop repository

https://github.com/mean-expert-official/loopback-sdk-builder/wiki/10.-SSL-Configuration

https://github.com/mean-expert-official/fireloop.io/wiki/Working-with-SSL

I will add this to the official docs ASAP.. I'm closing this issue now.

Cheers
Jon

**Comment by dmastag** [comment]
@jonathan-casarrubias thanks for the detailed instruction and glad its working for others. This was actually not working for me but I think its because I messed up my nginx settings so that the servers got mixed up 😞 

But anyway, I tried to replicate your settings to Apache2 and using url rewiring it was finally working.

In case this could help someone:
```
  RewriteEngine On
  RewriteCond %{HTTP:Connection} Upgrade [NC]
  RewriteRule /(.*) ws://localhost:3000/$1 [P,L]
```

Adding that did the job for apache.
